service-proxy.js

```diff
@@ -514,38 +520,71 @@ function refreshServicePeerPartially(serviceName, hostPort) {
     );
     var stop = (start + length) % workers.length;
 
-    self.logger.info('Refreshing service peer affinity', self.extendLogInfo({
-        serviceName: serviceName,
-        serviceHostPort: hostPort,
+    return {
+        start: start,
+        stop: stop,
         relayIndex: relayIndex,
         relays: relays,
         workers: workers,
-        start: start,
-        stop: stop,
         length: length
+    };
+};
+
+ServiceDispatchHandler.prototype.refreshServicePeerPartially =
+function refreshServicePeerPartially(serviceName, hostPort) {
+    var self = this;
+
+    var range = self.computePartialRange(serviceName, hostPort);
+
+    self.logger.info('Refreshing service peer affinity', self.extendLogInfo({
+        serviceName: serviceName,
+        serviceHostPort: hostPort,
+        relayIndex: range.relayIndex,
+        relays: range.relays,
+        workers: range.workers,
+        start: range.start,
+        stop: range.stop,
+        length: range.length
     }));
 
     // Open connections to affine peers
     var index;
     var peer;
-    if (start <= stop) { // ... start WITHIN stop ...
-        for (index = start; index < stop; index++) {
-            peer = self.getServicePeer(serviceName, workers[index]);
-            peer.connectTo();
+
+    // fully connected
+    if (range.start === range.stop) {
```

Comment:
I would say we need to thread modulo through here to distinguish fully connected from not connected at all, but I think we can rule out "not connected at all" on principle. :ok: